File: tests/vb/vb_collections_advanced/queue_operations.vb
' vybe-test: vb/vb_collections_advanced/queue_operations
' origin: languages/vb/tests/vb/test_vb_collections_advanced.rs

' Vybe test harness — Visual Basic.
'
' Real VB source alongside harness/go/check.go and harness/js/check.js, the way
' test262's assert.js is JavaScript.
'
' A test's verdict is its EXIT CODE. __Check prints its diagnostic BEFORE
' throwing: an uncaught exception surfaces as `RuntimeError: [object]`, which
' says nothing at all.

Module VybeCheck
    Sub __Check(got As String, want As String)
        If got <> want Then
            Console.WriteLine("FAIL: want [" & want & "] got [" & got & "]")
            Throw New Exception("assertion failed")
        End If
    End Sub
End Module

Module M
    Sub Main()
        Dim q As New Queue(Of String)
        q.Enqueue("first")
        q.Enqueue("second")
        q.Enqueue("third")
        __Check(CStr(q.Count), "3")
        __Check(CStr(q.Dequeue()), "first")
        __Check(CStr(q.Peek()), "second")
        __Check(CStr(q.Count), "2")
    End Sub
End Module

File: tests/csharp/csharp_collections/queue_operations.cs
// vybe-test: csharp/csharp_collections/queue_operations
// origin: languages/csharp/tests/csharp/test_csharp_collections.rs

using static __Harness;
using System.Collections.Generic;

var q = new Queue<string>();
q.Enqueue("first");
q.Enqueue("second");
q.Enqueue("third");
__P((q.Count).ToString());
__P((q.Dequeue()).ToString());
__P((q.Peek()).ToString());
__Check("3\nfirst\nsecond");

public static class __Harness {
    public static string __buf = "";
    public static void __P(string s) { __buf = __buf + s + "\n"; }
    public static void __Pr(string s) { __buf = __buf + s; }
    public static void __Check(string want) {
        if (__buf != want && __buf != want + "\n") {
            Console.WriteLine("FAIL: want [" + want + "] got [" + __buf + "]");
            throw new Exception("assertion failed");
        }
    }
}
